x-pack/legacy/plugins/fleet/dev_docs/fleet_agents_interactions_detailed.md

```diff
@@ -0,0 +1,46 @@
+# Fleet <-> Agent Interactions
+
+## Agent enrollment and checkin
+
+Fleet workflow:
+
+- an agent enroll to fleet using an enrollmentAPiKey
+- Every n seconds agent is polling the checkin API to send events and check for new configuration
+
+### Agent enrollment
+
+An agent must enroll using the REST Api provided by fleet.
+When an agent enroll Fleet:
+
+- verify the API Key is a valid ES API key
+- retrieve the Saved Object (SO) associated to this api key id (this SO contains the configuration|policy id)
+- create an ES ApiKey unique to the agent for accessing kibana during checkin
+- create an ES ApiKey per output to send logs and metrics to the output
+- Save the new agent in a SO with keys encrypted inside the agent SO object
```

Comment:
Do we need to store the Agent's ES API Keys Key in the agent saved-object? Shouldn't the API Key ID be sufficient? If so, I don't think we'd want to encrypt the API Key ID. If we do, we won't be able to search on it.

Reply:
We need the API Key as we need it to send it to the agent.

Reply:
If it's only being sent to the agent when they first enroll, it doesn't need to be stored for subsequent access, right?

Reply:
The keys are part of the config. So each time a config is updated, the keys are sent down again. It could be that the first config sent down only contains 1 output, the second config contains 2 outputs so 2 api keys.